### PR TITLE
DM-21043: Qserv log diet: use named context for query ID

### PR DESCRIPTION
**Pull request by andy-slac**
Implement scope guard for logging MDC.

This small addition to log package is needed to support scope-based MDC control in qserv. It could be useful for other clients too.

### CODE REVIEW

File: include/lsst/log/Log.h [hunk 1]
```diff
@@ -104,6 +104,18 @@
...
 
+/**
+  * @def LOG_MDC_SCOPE(key, value)
+  * Places a KEY/VALUE pair in the Mapped Diagnostic Context (MDC) for the
+  * current thread, and restores previous KEY/VALUE on exit from a scope.
+  *
+  * @param key    Unique key.
+  * @param value  String value.
+  */
+#define LOG_CONCAT_IMPL(a, b) a ## b
```
Comment (ktlim) [inline include/lsst/log/Log.h]: A comment here about the trickiness of getting `__COUNTER__` evaluated would be nice.

File: doc/mainpage.dox [hunk 2]
```diff
@@ -352,13 +352,14 @@ For example, a session id may be included by using the following formatting stri
 
     "%d [%t] %-5p %c{2} (%X{session}) %m%n"
 
-or alternatively using `%%X` format code without additional key to display full MDC:
+or alternatively using `%X` format code without additional key to display full MDC:
```
Comment (ktlim) [inline doc/mainpage.dox]: I think the double-`%` is necessary for Doxygen.  It appears correct in http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/log.html

File: include/lsst/log/Log.h (hunk 1 above)
Comment (andy-slac) [reply include/lsst/log/Log.h]: Hmm, this is a standard C macro mess/trickiness, Did not think it needs special comment after decades of history 🙂 I'll say something wise.

File: doc/mainpage.dox (hunk 2 above)
Comment (andy-slac) [reply doc/mainpage.dox]: I guess I was confused by other cases of single percent. I guess they are all wrong and now I need to fix all of them. Thank you, K-T :sigh:

Comment (andy-slac) [reply doc/mainpage.dox]: That percent thing is odd, I can't figure out exact rules but I think I got it right after adding/removing double-% in few places and looking at resulting output.

**Pull request merged**